### PR TITLE
Adding comments and reasons to Restart-Computer/Stop-Computer

### LINKED ISSUE #4857: Restart-Computer and Stop-Computer should take custom message like shutdown.exe
When using stop-computer or restart-computer, one should be able to log a reason code and custom message similar to the way that shutdown.exe with /d P:0:1 and /c "Custom Message".  This would allow you to stay in powershell without needing to call out to win32 executable.  See similar concern:
https://stackoverflow.com/questions/15656106/restarting-pcs-and-writing-to-the-event-log-with-a-custom-message

### PR DESCRIPTION
**Pull request by lukepayne**
## PR Summary

This change adds the ability to specify a comment as well as reason codes
for a reboot or shutdown of the computer. Also added is the option to
specify a timeout before the reboot/shutdown occurs.

fixes #4857
## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [X] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by vexx32** [comment]
@lukepayne if you put something like "fixes #4857" in the PR description, then the issue will be closed when the maintainers merge the PR. Just saves them a few moments extra work 😉

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs [hunk 1]
```diff
@@ -173,6 +173,44 @@ public class RestartComputerCommand : PSCmdlet, IDisposable
...
+        private const uint Reconfigure = 0x00000004;
+        private const uint SecurityIssue = 0x00000013;
+        private const uint SecurityPatch = 0x00000012;
+        private const uint SecurityPatchUninstallation = 0x00000018;
+        private const uint ServicePack = 0x00000010;
+        private const uint ServicePackUninstallation = 0x00000016;
+        private const uint TerminalServices = 0x00000020;
+        private const uint Unstable = 0x00000006;
+        private const uint Upgrade = 0x00000003;
+        private const uint WMI = 0x00000015;
```
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: There are quite a few reason odes available which would indicate to me that this would be much better off as an `enum` for cleanliness and clarity; much easier for users to discover as well, if the parameter is typed with the enum type!

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs [hunk 2]
```diff
@@ -223,6 +261,74 @@ public class RestartComputerCommand : PSCmdlet, IDisposable
...
+            "SecurityIssue",
+            "SecurityPatch",
+            "SecurityPatchUninstallation",
+            "ServicePack",
+            "ServicePackUninstallation",
+            "TerminalServices",
+            "Unstable",
+            "Upgrade",
+            "WMI")]
+        public string MinorReason { get; set; }
```
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Per my other comment, you can skip ValidateSet entirely and you could opt to either have two enums (one for major and one for minor reason) or you could opt to roll it all into a single enum and parameter, depending on which is appropriate.

If multiple codes can be specified, tag the enum(s) with the `[Flags]` attribute to permit this.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs [hunk 3]
```diff
@@ -800,7 +906,7 @@ protected override void BeginProcessing()
                     case WaitForServiceTypes.WinRM:
                         break;
                     case WaitForServiceTypes.PowerShell:
-                        _powershell = System.Management.Automation.PowerShell.Create();
+                        _powershell = global::System.Management.Automation.PowerShell.Create();
```
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Is this change necessary & what does it accomplish here? Never seen this sort of syntax in c# before, and I don't know if it's used anywhere else in the repo here.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs [hunk 4]
```diff
@@ -1102,6 +1320,43 @@ public sealed class StopComputerCommand : PSCmdlet, IDisposable
...
+        private const uint Reconfigure = 0x00000004;
+        private const uint SecurityIssue = 0x00000013;
+        private const uint SecurityPatch = 0x00000012;
+        private const uint SecurityPatchUninstallation = 0x00000018;
+        private const uint ServicePack = 0x00000010;
+        private const uint ServicePackUninstallation = 0x00000016;
+        private const uint TerminalServices = 0x00000020;
+        private const uint Unstable = 0x00000006;
+        private const uint Upgrade = 0x00000003;
+        private const uint WMI = 0x00000015;
```
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Same as above; it looks like you could roll most or all of these into a single enum that is shared between both commands.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs [hunk 5]
```diff
@@ -2027,21 +2462,21 @@ internal static bool SkipSystemRestoreOperationForARMPlatform(PSCmdlet cmdlet)
...
-        /// <param name="flags">Win32Shutdown flags</param>
+        /// <param name="flags">Win32ShutdownTracker flags</param>
         /// <param name="credential">Optional credential</param>
         /// <param name="authentication">Optional authentication</param>
         /// <param name="formatErrorMessage">Error message format string that takes two parameters</param>
         /// <param name="ErrorFQEID">Fully qualified error Id</param>
         /// <param name="cancelToken">Cancel token</param>
         /// <returns>True on success</returns>
-        internal static bool InvokeWin32ShutdownUsingWsman(
+        internal static bool InvokeWin32ShutdownTrackerUsingWsman(
```
Comment (vexx32) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: I don't think it's a good idea to change the method name itself. If a new overload or a new method that does what you need is necessary, that's probably better, unless you're absolutely sure that nothing else calls this method at any point. It's flagged as `internal`, so potentially anything in the same namespace could call it, though I'm not sure if anything currently does.

Comment (lukepayne) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: There's no specific reason why it needs to be changed, it's just to match the name of the new CimOperatingSystemShutdownMethod. I can revert to the previous name as I don't think an overload is necessary.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs (hunk 3 above)
Comment (lukepayne) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: This forces the call to search in the global namespace, otherwise the 'System' prefix grabs the local System variable. However, this likely won't be an issue after the the reason codes are converted to enums.

**Comment by lukepayne** [comment]
> @lukepayne if you put something like "fixes #4857" in the PR description, then the issue will be closed when the maintainers merge the PR. Just saves them a few moments extra work 😉

Thanks for the tip! I'll add that in :)

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs (hunk 5 above)
Comment (vexx32) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Gotcha. Yeah, if you can check the references to that method to see if anything else happened to also use it that would be best.

Otherwise I'll be more than happy to do a quick check in a bit just to make sure. 😆

Comment (vexx32) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Just did a quick check, and it doesn't look like anything else is using this method, so it's probably fine either way. 😄

Comment (lukepayne) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Oh great, thanks. I might leave it as is then to match the new shutdown method name 😀

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs [hunk 6]
```diff
@@ -173,6 +173,44 @@ public class RestartComputerCommand : PSCmdlet, IDisposable
         private const string DefaultParameterSet = "DefaultSet";
         private const int forcedReboot = 6; // see https://msdn.microsoft.com/library/aa394058(v=vs.85).aspx
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Please update the link. (Now it is redirected).
And add doc link for new constants you are added below.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs [hunk 7]
```diff
@@ -223,6 +261,74 @@ public class RestartComputerCommand : PSCmdlet, IDisposable
         [Alias("f")]
         public SwitchParameter Force { get; set; }
 
+        /// <summary>
+        /// Provide a comment for rebooting the computer.
+        /// </summary>
+        [Parameter(ParameterSetName = DefaultParameterSet)]
+        public string Comment { get; set; } = "";
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Please use `string.Empty`.

Style issue: all public comments for properties with `{ get; set; }` should start with "Gets or sets"; (line 265)

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs [hunk 8]
```diff
@@ -223,6 +261,74 @@ public class RestartComputerCommand : PSCmdlet, IDisposable
...
+        [Parameter(ParameterSetName = DefaultParameterSet)]
+        [ValidateSet(
+            "Application",
+            "Hardware",
+            "OperatingSystem",
+            "Other",
+            "Power",
+            "Software",
+            "System")] 
+        public string MajorReason { get; set; }
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: It should enum too.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs (hunk 2 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Seems we should review the list. How we can call the cmdlet if "BlueScreen"?

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs [hunk 9]
```diff
@@ -223,6 +261,74 @@ public class RestartComputerCommand : PSCmdlet, IDisposable
...
+        [ValidateRange(0, UInt32.MaxValue)]
+        public UInt32 RestartDelay
+        {
+            get { return _restartDelay; }
+            set
+            {
+                _restartDelay = value;
+            }
+        }
+        private UInt32 _restartDelay = 0;
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Please remove and use auto-property `public UInt32 RestartDelay { get; set; }`.
No need to init - default is 0.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs [hunk 10]
```diff
@@ -1028,7 +1246,7 @@ protected override void ProcessRecord()
                             {
                                 _status = ComputerResources.RestartComplete;
                                 WriteProgress(_indicator[indicatorIndex % 4] + _activity, _status, 100, ProgressRecordType.Completed);
-                                _timer.Change(System.Threading.Timeout.Infinite, System.Threading.Timeout.Infinite);
+                                _timer.Change(global::System.Threading.Timeout.Infinite, global::System.Threading.Timeout.Infinite);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: `Win32ShutdownTracker` supports timeout so we could remove the code. /cc @SteveL-MSFT

**Review by iSazonov — COMMENTED** [review]
We shouldn't duplicate code (dry principle) in both cmdlets.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs (hunk 2 above)
Comment (lukepayne) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: I tried to go for parity with what the Win32ShutdownTracker supports, but it makes sense to remove some of the illogical ones. Do you have suggestions on what should be removed?

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: I'd look options in Windows restart/shootdown dialog box.

Comment (lukepayne) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: The options from the Server shutdown dialog include the following options. It seems a bit restrictive given what the API supports.

image
image

Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: > It seems a bit restrictive given what the API supports.

because not all options make sense in interactive mode.

Comment (lukepayne) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Correct. So I would suggest not whittling it down to this list.

Comment (lukepayne) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Correct. So I would suggest not whittling it down to this list.

**Comment by lukepayne** [comment]
I'm hoping a comment is sufficient to keep this open for now. I intend to finish and commit once I'm back from holidays. Thanks!

**Comment by kvprasoon** [comment]
@lukepayne or anyone working on this ?
